integration-testing/src/tools/network.py

```diff
@@ -0,0 +1,40 @@
+from contextlib import contextmanager
+from tools.wait import wait_for, has_peers, node_started
+import logging
+from tools.rnode import create_peer_nodes
+
+
+class RChain:
+    def __init__(self, network, bootstrap, peers):
+        self.network = network
+        self.bootstrap = bootstrap
+        self.peers = peers
+        self.nodes = [bootstrap] + peers
+
+
+@contextmanager
+def start_network(config, docker, bootstrap, validators_data, allowed_peers=None):
+    logging.debug(f"Docker network = {bootstrap.network}")
+
+    bonds_file, _, peer_keys = validators_data
+    peers = create_peer_nodes(docker, bootstrap, bootstrap.network, bonds_file, peer_keys, config.rnode_timeout, allowed_peers)
+
+    yield RChain(network = bootstrap.network, bootstrap = bootstrap, peers = peers)
+
+    for peer in peers:
+        peer.cleanup()
+
+
+def wait_for_started_network(node_startup_timeout, network):
+    for peer in network.peers:
+        wait_for(node_started(peer), node_startup_timeout, f"Peer {peer.name} did not start correctly.")
```

Comment:
I'm strongly against using callback in this way.  It is much more readable and debuggable to have a explicit loop.  Higher order functions in Python are more problematic than in a statically-typed language.

Reply:
wait_for is a generic function. It waits for any condition. 

If you want a specific function for a certain condition you can implement it on top of wait_for and a couple of basic predicates

Reply:
What you're saying is crystal clear.  The problem is with debugging such a code.  A test will fail eventually.  When it fails, the stack trace should ideally tell the expectation that failed.  With `wait_for()` the stack trace will be deeper, names more cryptic and the reason for failing more obscure.  Trust me, let's not take this path.